File: julia/recursed/collatzRecur.jl
#!/usr/bin/julia

function collatz(value)
	if value % 2 == 0
		return collatz(value /= 2) + 1
    elseif value % 2 == 1 && value > 1
        return collatz(value * 3 + 1) + 1
    else
		return 0
	end
end

function bubblesort!(arr::AbstractVector)
    for _ in 2:length(arr), j in 1:length(arr)-1
        if arr[j] > arr[j+1]
            arr[j], arr[j+1] = arr[j+1], arr[j]
        end
    end
    return arr
end

numbers = [0,0,0,0,0,0,0,0,0,0]
sequences = [0,0,0,0,0,0,0,0,0,0]
maxIndex=4999999999

numbers=bubblesort!(numbers)
sequences=bubblesort!(sequences)


for i in 1:maxIndex
    ic = 0
    ic=collatz(i)
    if ic > sequences[1]
        global sequences[1]=ic
        global numbers[1]=i
    elseif ic == sequences[1]
        if i < numbers[1]
            global numbers[1]=i
        end
    end
    global numbers=bubblesort!(numbers)
    global sequences=bubblesort!(sequences)  
end 

println("Numbers: ",numbers)
println("Sequences: ",sequences)


exit(0)

File: go/recursed/collatzRecur.go
package main
import "fmt"

func collatz(x int) int {
    var count int = 0
    for {
        if x == 1 {
            break
        } else if x%2 == 0 {
            x = x / 2
            return collatz(x) + 1
        } else {
            x = 3 * x + 1
            return collatz(x) + 1
        }
        count++
    }
    return count
}

func bubblesort(a []int) {
    for itemCount := len(a) - 1; ; itemCount-- {
        hasChanged := false
        for index := 0; index < itemCount; index++ {
            if a[index] > a[index+1] {
                a[index], a[index+1] = a[index+1], a[index]
                hasChanged = true
            }
        }
        if hasChanged == false {
            break
        }
    }
}

func main() {
	var ic int 
    var i int
    var maxIndex int
    numbers := []int{0,0,0,0,0,0,0,0,0,0}
    sequences := []int{0,0,0,0,0,0,0,0,0,0}
    maxIndex=5000000000

    bubblesort(numbers)
    bubblesort(sequences)
    i=1
    for i < maxIndex {
        ic=collatz(i)
        if ic > sequences[0] {
            sequences[0]=ic
            numbers[0]=i
        } else if ic == sequences[0] {
            if i < numbers[0] {
                numbers[0]=i
            }
        }
        bubblesort(numbers)
        bubblesort(sequences)
        i=i+1
    }

    fmt.Println("Numbers: ", numbers)
    fmt.Println("Sequences: ", sequences)

}